### PR TITLE
[Search RM Pagination] case insensitive operators and parameter validation

### PR DESCRIPTION
**Pull request by ankit-db**
## What changes are proposed in this pull request?

Mani highlighted some slight issues with the initial PR for the DB implementation of pagination for `search-registered-models`. This PR fixes those issues.

## How is this patch tested?

Unit tests

### Is this a user-facing change?

- [x] No. You can skip the rest of this section.
- [ ] Yes. Give a description of this change to be included in the release notes for MLflow users.

(Details in 1-2 sentences. You can just refer to another PR with a description if this PR is part of a larger change.)

### What component(s), interfaces, languages, and integrations does this PR affect?
Components 
- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for
Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: Local serving, model deployment tools, spark UDFs
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface 
- [ ] `area/uiux`: Front-end, user experience, JavaScript, plotting
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language 
- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients

Integrations
- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations

### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [ ] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

### CODE REVIEW

File: mlflow/store/model_registry/sqlalchemy_store.py [hunk 1]
```diff
@@ -258,23 +258,20 @@ def compute_next_token(current_size):
...
                                       'match (ILIKE). Input filter string: %s' % filter_string,
                                       error_code=INVALID_PARAMETER_VALUE)
-            if filter_dict["key"] == "name":
-                if filter_dict["comparator"] == "LIKE":
-                    conditions = [SqlRegisteredModel.name.like(filter_dict["value"])]
-                elif filter_dict["comparator"] == "ILIKE":
-                    conditions = [SqlRegisteredModel.name.ilike(filter_dict["value"])]
-                else:
-                    conditions = [SqlRegisteredModel.name == filter_dict["value"]]
+            if comparator == "LIKE":
```
Comment (mparkhe) [inline mlflow/store/model_registry/sqlalchemy_store.py:269]: oh! do we have constants from `SearchUtils` for each of these operators? Might be a good idea to add those.

File: mlflow/utils/search_utils.py [hunk 2]
```diff
@@ -488,5 +490,26 @@ def parse_filter_for_model_registry(cls, filter_string):
...
+    @classmethod
+    def parse_filter_for_model_versions(cls, filter_string):
+        statement = cls._parse_filter_for_model_registry(filter_string)
+        if statement:
+            return [cls._get_comparison_for_model_registry(
+                si,
+                cls.VALID_SEARCH_KEYS_FOR_MODEL_VERSIONS)
+                    for si in statement.tokens if isinstance(si, Comparison)]
+        else:
+            return []
```
Comment (mparkhe) [inline mlflow/utils/search_utils.py:504]: why not push `cls.VALID_SEARCH_KEYS_FOR_MODEL_VERSIONS` into `_parse_filter_for_model_registry` method and [DRY this code out](https://en.wikipedia.org/wiki/Don%27t_repeat_yourself), like

```
@classmethod
def _parse_filter_for_model_registry(cls, filter_string, valid_search_keys):
  # ...
  # existing unchanged code
  # ...
  return [cls._get_comparison_for_model_registry(si, valid_search_keys)
          for si in statement.tokens if isinstance(si, Comparison)]	  

@classmethod
def parse_filter_for_model_versions(cls, filter_string):
  return cls._parse_filter_for_model_registry(filter_string, cls.VALID_SEARCH_KEYS_FOR_MODEL_VERSIONS)

@classmethod
def parse_filter_for_registered_models(cls, filter_string):
  return cls._parse_filter_for_model_registry(filter_string, cls.VALID_SEARCH_KEYS_FOR_REGISTERED_MODELS)
```

File: tests/store/model_registry/test_sqlalchemy_store.py [hunk 3]
```diff
@@ -502,6 +506,10 @@ def test_search_registered_models(self):
         rms, _ = self._search_registered_models(f"name ILIKE '{prefix + 'cats'}%'")
         self.assertEqual(rms, [])
 
+        # confirm that ILIKE is not case-sensitive
+        rms, _ = self._search_registered_models(f"name ilike '%RM4a'")
```
Comment (mparkhe) [inline tests/store/model_registry/test_sqlalchemy_store.py:510]: also add a negative test for case-insensitive operator : searching with filter `"name iLike '%RM17%'"` should yield empty result.

Comment (ankitmathur-db) [reply tests/store/model_registry/test_sqlalchemy_store.py:510]: Done!

File: mlflow/utils/search_utils.py (hunk 2 above)
Comment (ankitmathur-db) [reply mlflow/utils/search_utils.py:504]: Yeah, this makes sense - thanks for the suggestion!

File: mlflow/store/model_registry/sqlalchemy_store.py (hunk 1 above)
Comment (ankitmathur-db) [reply mlflow/store/model_registry/sqlalchemy_store.py:269]: Added!

File: tests/store/model_registry/test_sqlalchemy_store.py [hunk 4]
```diff
@@ -502,14 +509,21 @@ def test_search_registered_models(self):
...
+
         # cannot search by invalid comparator types
         with self.assertRaises(MlflowException) as exception_context:
             self._search_registered_models("name!=something")
         assert exception_context.exception.error_code == ErrorCode.Name(INVALID_PARAMETER_VALUE)
 
         # cannot search by run_id
         with self.assertRaises(MlflowException) as exception_context:
-            self._search_registered_models("run_id='%s'" % "somerunID")
+            print(self._search_registered_models("run_id='%s'" % "somerunID"))
```
Comment (mparkhe) [inline tests/store/model_registry/test_sqlalchemy_store.py:526]: remove `print`

**Review by mparkhe — COMMENTED** [review]
remove `print` -- otherwise looks good.

**Review by mparkhe — APPROVED** [review]

**Pull request merged**